File: src/lib/serpapi.ts
/**
 * SerpApi Google Flights client.
 *
 * Source of the low/typical/high classification. SerpApi proxies Google
 * Flights and exposes `price_insights`:
 *
 *   { lowest_price, price_level, typical_price_range: [lo, hi],
 *     price_history: [[unix_ts, price], ...] }
 */

const ENDPOINT = "https://serpapi.com/search.json";

export class ProviderError extends Error {
  override name = "ProviderError";
}

export interface Offer {
  price: number;
  airline: string;
  flightNumbers: string;
  stops: number;
  totalDuration: number | null;
  departureTime: string | null;
  arrivalTime: string | null;
}

export interface FetchResult {
  lowestPrice: number | null;
  avgPrice: number | null;
  medianPrice: number | null;
  nOffers: number;
  priceLevel: string | null;
  typicalLow: number | null;
  typicalHigh: number | null;
  /** Google's own daily history, already converted to YYYY-MM-DD. */
  history: { date: string; price: number }[];
  offers: Offer[];
}

export interface FetchParams {
  origin: string;
  destination: string;
  outboundDate: string;
  returnDate: string | null;
  adults?: number;
  currency?: string;
  travelClass?: number;
  stops?: number;
  apiKey?: string;
  signal?: AbortSignal;
}

interface RawLeg {
  departure_airport?: { name?: string; id?: string; time?: string };
  arrival_airport?: { name?: string; id?: string; time?: string };
  airline?: string;
  flight_number?: string;
}

interface RawItinerary {
  flights?: RawLeg[];
  layovers?: { duration?: number; id?: string }[];
  total_duration?: number;
  price?: number;
}

function parseOffer(item: RawItinerary): Offer | null {
  if (typeof item.price !== "number") return null;
  const legs = item.flights ?? [];
  const first = legs[0];
  const last = legs[legs.length - 1];
  const airlines = [...new Set(legs.map((l) => l.airline).filter(Boolean))] as string[];

  return {
    price: item.price,
    airline: airlines.join(", "),
    flightNumbers: legs.map((l) => l.flight_number ?? "").filter(Boolean).join(", "),
    stops: Math.max(legs.length - 1, 0),
    totalDuration: item.total_duration ?? null,
    departureTime: first?.departure_airport?.time ?? null,
    arrivalTime: last?.arrival_airport?.time ?? null,
  };
}

function median(values: number[]): number {
  const sorted = [...values].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 === 0 ? (sorted[mid - 1]! + sorted[mid]!) / 2 : sorted[mid]!;
}

export async function fetchRoute(params: FetchParams): Promise<FetchResult> {
  const apiKey = params.apiKey ?? process.env.SERPAPI_KEY;
  if (!apiKey) throw new ProviderError("SERPAPI_KEY is not set");

  const query = new URLSearchParams({
    engine: "google_flights",
    api_key: apiKey,
    departure_id: params.origin,
    arrival_id: params.destination,
    outbound_date: params.outboundDate,
    type: params.returnDate ? "1" : "2",
    currency: params.currency ?? "USD",
    hl: "en",
    gl: "us",
    adults: String(params.adults ?? 1),
    travel_class: String(params.travelClass ?? 1),
    stops: String(params.stops ?? 0),
    deep_search: "true",
  });
  if (params.returnDate) query.set("return_date", params.returnDate);

  let response: Response;
  try {
    response = await fetch(`${ENDPOINT}?${query}`, {
      signal: params.signal,
      headers: { "User-Agent": "flight-tracker/1.0" },
    });
  } catch (cause) {
    throw new ProviderError(`Could not reach SerpApi: ${(cause as Error).message}`);
  }

  if (!response.ok) {
    const body = (await response.text()).slice(0, 300);
    throw new ProviderError(`SerpApi HTTP ${response.status}: ${body}`);
  }

  const payload = (await response.json()) as Record<string, unknown>;
  if (payload.error) throw new ProviderError(`SerpApi error: ${String(payload.error)}`);

  const best = (payload.best_flights as RawItinerary[]) ?? [];
  const other = (payload.other_flights as RawItinerary[]) ?? [];
  const offers = [...best, ...other]
    .map(parseOffer)
    .filter((o): o is Offer => o !== null)
    .sort((a, b) => a.price - b.price);

  const prices = offers.map((o) => o.price);
  const insights = (payload.price_insights as Record<string, unknown> | undefined) ?? {};
  const typical = (insights.typical_price_range as [number, number] | undefined) ?? [null, null];

  if (prices.length === 0 && Object.keys(insights).length === 0) {
    throw new ProviderError(
      `No flights and no price insights for ${params.origin}->${params.destination} ` +
        `on ${params.outboundDate}. Check the airport codes and that the date is inside ` +
        `the airlines' bookable window (~11 months).`,
    );
  }

  const rawHistory = (insights.price_history as [number, number][] | undefined) ?? [];
  const history: { date: string; price: number }[] = [];
  for (const entry of rawHistory) {
    if (!Array.isArray(entry) || entry.length !== 2) continue;
    const [ts, price] = entry;
    if (typeof ts !== "number" || typeof price !== "number") continue;
    const date = new Date(ts * 1000).toISOString().slice(0, 10);
    history.push({ date, price: Math.round(price) });
  }

  return {
    lowestPrice:
      (insights.lowest_price as number | undefined) ?? (prices.length ? prices[0]! : null),
    avgPrice: prices.length
      ? Math.round((prices.reduce((a, b) => a + b, 0) / prices.length) * 100) / 100
      : null,
    medianPrice: prices.length ? Math.round(median(prices) * 100) / 100 : null,
    nOffers: prices.length,
    priceLevel: (insights.price_level as string | undefined) ?? null,
    typicalLow: typical[0] ?? null,
    typicalHigh: typical[1] ?? null,
    history,
    offers: offers.slice(0, 8),
  };
}

File: python/flighttracker/providers/serpapi.py
"""SerpApi Google Flights client.

This is the source of the low/typical/high classification. SerpApi proxies
Google Flights and exposes `price_insights`:

    {
      "lowest_price": 612,
      "price_level": "low" | "typical" | "high",
      "typical_price_range": [570, 1050],
      "price_history": [[unix_ts, price], ...]
    }
"""

from __future__ import annotations

import json
import statistics
import urllib.error
import urllib.parse
import urllib.request
from datetime import datetime, timezone
from pathlib import Path

from .. import config as config_module
from ..config import Config, Trip, require_key

ENDPOINT = "https://serpapi.com/search.json"
KEY_URL = "https://serpapi.com/manage-api-key"


class ProviderError(Exception):
    pass


def _get(params: dict, timeout: int = 60) -> dict:
    url = f"{ENDPOINT}?{urllib.parse.urlencode(params)}"
    req = urllib.request.Request(url, headers={"User-Agent": "flight-tracker/1.0"})
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            payload = json.loads(resp.read().decode("utf-8"))
    except urllib.error.HTTPError as exc:
        body = exc.read().decode("utf-8", errors="replace")[:500]
        raise ProviderError(f"SerpApi HTTP {exc.code}: {body}") from exc
    except urllib.error.URLError as exc:
        raise ProviderError(f"Could not reach SerpApi: {exc.reason}") from exc

    if "error" in payload:
        raise ProviderError(f"SerpApi returned an error: {payload['error']}")
    return payload


def _parse_offer(item: dict, is_best: bool) -> dict:
    legs = item.get("flights", [])
    first, last = (legs[0], legs[-1]) if legs else ({}, {})
    return {
        "price": item.get("price"),
        "airline": ", ".join(sorted({leg.get("airline", "") for leg in legs if leg.get("airline")})),
        "flight_numbers": ", ".join(leg.get("flight_number", "") for leg in legs),
        "stops": max(len(legs) - 1, 0),
        "total_duration": item.get("total_duration"),
        "departure_time": (first.get("departure_airport") or {}).get("time"),
        "arrival_time": (last.get("arrival_airport") or {}).get("time"),
        "layovers": [
            {"airport": lay.get("id"), "minutes": lay.get("duration")}
            for lay in item.get("layovers", [])
        ],
        "is_best": is_best,
        "booking_token": item.get("booking_token"),
    }


def fetch(trip: Trip, cfg: Config, save_raw: bool = True) -> dict:
    """Poll Google Flights via SerpApi and normalize into a snapshot dict."""
    api_key = require_key("SERPAPI_KEY", KEY_URL)

    params = {
        "engine": "google_flights",
        "api_key": api_key,
        "departure_id": trip.origin,
        "arrival_id": trip.destination,
        "outbound_date": trip.outbound_date,
        "type": trip.trip_type,
        "currency": cfg.currency,
        "hl": "en",
        "gl": "us",
        "travel_class": cfg.travel_class,
        "stops": cfg.stops,
        "deep_search": "true",  # match what google.com/travel/flights actually shows
        **{k: v for k, v in cfg.passengers.items() if v},
    }
    if trip.return_date:
        params["return_date"] = trip.return_date

    payload = _get(params)

    raw_path = None
    if save_raw:
        raw_dir = config_module.RAW_DIR
        raw_dir.mkdir(parents=True, exist_ok=True)
        stamp = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
        target = raw_dir / f"{trip.route_id}_{stamp}.json"
        target.write_text(json.dumps(payload, indent=2))
        raw_path = str(Path(target).relative_to(raw_dir.parents[1]))

    best = payload.get("best_flights") or []
    other = payload.get("other_flights") or []
    offers = [_parse_offer(i, True) for i in best] + [_parse_offer(i, False) for i in other]
    prices = [o["price"] for o in offers if isinstance(o.get("price"), (int, float))]

    insights = payload.get("price_insights") or {}
    typical = insights.get("typical_price_range") or [None, None]

    if not prices and not insights:
        raise ProviderError(
            f"No flights and no price insights returned for {trip.route_id}. "
            "Check the airport codes and that the dates are within the airlines' "
            "bookable window (usually ~11 months out)."
        )

    return {
        "provider": "serpapi_google_flights",
        "route_id": trip.route_id,
        "trip_id": trip.id,
        "origin": trip.origin,
        "destination": trip.destination,
        "outbound_date": trip.outbound_date,
        "return_date": trip.return_date,
        "currency": cfg.currency,
        "lowest_price": insights.get("lowest_price") or (min(prices) if prices else None),
        "avg_price": round(statistics.fmean(prices), 2) if prices else None,
        "median_price": round(statistics.median(prices), 2) if prices else None,
        "n_offers": len(prices),
        "price_level": insights.get("price_level"),
        "typical_low": typical[0] if len(typical) == 2 else None,
        "typical_high": typical[1] if len(typical) == 2 else None,
        "price_history": insights.get("price_history") or [],
        "offers": offers,
        "raw_path": raw_path,
    }


def history_points(snapshot: dict) -> list[tuple[str, int]]:
    """Convert Google's price_history [[unix_ts, price], ...] to (date, price)."""
    points = []
    for entry in snapshot.get("price_history", []):
        if not isinstance(entry, (list, tuple)) or len(entry) != 2:
            continue
        ts, price = entry
        try:
            date = datetime.fromtimestamp(int(ts), tz=timezone.utc).date().isoformat()
            points.append((date, int(price)))
        except (ValueError, TypeError, OSError):
            continue
    return points
